source/common/quic/envoy_quic_server_stream.cc

```diff
@@ -82,31 +83,37 @@ void EnvoyQuicServerStream::encodeData(Buffer::Instance& data, bool end_stream)
   ASSERT(!local_end_stream_);
   local_end_stream_ = end_stream;
   SendBufferMonitor::ScopedWatermarkBufferUpdater updater(this, this);
-  Buffer::RawSliceVector raw_slices = data.getRawSlices();
-  absl::InlinedVector<quiche::QuicheMemSlice, 4> quic_slices;
-  quic_slices.reserve(raw_slices.size());
-  for (auto& slice : raw_slices) {
-    ASSERT(slice.len_ != 0);
-    // Move each slice into a stand-alone buffer.
-    // TODO(danzh): investigate the cost of allocating one buffer per slice.
-    // If it turns out to be expensive, add a new function to free data in the middle in buffer
-    // interface and re-design QuicheMemSliceImpl.
-    quic_slices.emplace_back(quiche::QuicheMemSlice::InPlace(), data, slice.len_);
-  }
-  quic::QuicConsumedData result{0, false};
-  absl::Span<quiche::QuicheMemSlice> span(quic_slices);
-  {
-    IncrementalBytesSentTracker tracker(*this, *mutableBytesMeter(), false);
-    result = WriteBodySlices(span, end_stream);
-    stats_gatherer_->addBytesSent(result.bytes_consumed, end_stream);
-  }
-  // QUIC stream must take all.
-  if (result.bytes_consumed == 0 && has_data) {
-    IS_ENVOY_BUG(fmt::format("Send buffer didn't take all the data. Stream is write {} with {} "
-                             "bytes in send buffer. Current write was rejected.",
-                             write_side_closed() ? "closed" : "open", BufferedDataBytes()));
+  if (capsule_protocol_handler_.usingCapsuleProtocol() &&
+      !capsule_protocol_handler_.encodeCapsule(data.toString())) {
+    // TODO(jeongseokson): Check if resetting the stream is a right way to handle the error.
```

Comment:
See other comments, I don't think this failure can happen in practice?

Reply:
Done.